src/dataregistry/registrar/dataset.py

```diff
@@ -7,403 +7,481 @@
+			valid_status = 0
+
+		# Case where use is overwriting the dateset `creation_date`
+		if creation_date:
+			ds_creation_date = creation_date
+
+		# Copy was successful, update the entry with dataset metadata
+		with self._engine.connect() as conn:
+			update_stmt = (
+				update(dataset_table)
+				.where(dataset_table.c.dataset_id == prim_key)
+				.values(
+					data_org=dataset_organization,
+					nfiles=num_files,
+					total_disk_space=total_size / 1024 / 1024,
+					creation_date=ds_creation_date,
+					status=set_dataset_status(values["status"], valid=True),
+				)
+			)
+			conn.execute(update_stmt)
+			conn.commit()
+
+		return prim_key, execution_id
+
+	def _handle_data(self, relative_path, old_location, owner, owner_type, verbose):
+		"""
+		Find characteristics of dataset (i.e., is it a file or directory, how
+		many files and total disk space of the dataset).
+
+		If old_location is not None, copy the dataset files and directories
+		into the data registry.
+
+		Parameters
+		----------
+		relative_path : str
+			Relative path of dataset in the data registry
+		old_location : str
+			Location of data (if not already in the data registry root)
+			Data will be copied from this location
+		owner : str
+			Owner of the dataset
+		owner_type : str
+			Owner type of the dataset
+		verbose : bool
+			True for extra output
+
+		Returns
+		-------
+		dataset_organization : str
+			"file", "directory", or "dummy"
+		num_files : int
+			Total number of files making up dataset
+		total_size : float
+			Total disk space of dataset in bytes
+		ds_creation_date : datetime
+			When file or directory was created
+		"""
+
+		# Get destination directory in data registry.
+		dest = _form_dataset_path(
+			owner_type,
+			owner,
+			relative_path,
+			schema=self._schema,
+			root_dir=self._root_dir,
+		)
+
+		# Is the data already on location, or coming from somewhere new?
+		if old_location:
+			loc = old_location
+		else:
+			loc = dest
+
+		# Get metadata on dataset.
+		if os.path.isfile(loc):
+			dataset_organization = "file"
+		elif os.path.isdir(loc):
+			dataset_organization = "directory"
+		else:
+			raise FileNotFoundError(f"Dataset {loc} not found")
+
+		if verbose:
+			tic = time.time()
+			print("Collecting metadata...", end="")
+
+		ds_creation_date = datetime.fromtimestamp(os.path.getctime(loc))
+
+		if dataset_organization == "directory":
+			num_files, total_size = get_directory_info(loc)
+		else:
+			num_files = 1
+			total_size = os.path.getsize(loc)
+		if verbose:
+			print(f"took {time.time()-tic:.2f}s")
+
+		# Copy data into data registry
+		if old_location:
+			if verbose:
+				tic = time.time()
+				print(
+					f"Copying {num_files} files ({total_size/1024/1024:.2f} Mb)...",
+					end="",
+				)
+			_copy_data(dataset_organization, old_location, dest)
+			if verbose:
+				print(f"took {time.time()-tic:.2f}")
+
+		return dataset_organization, num_files, total_size, ds_creation_date
+
+	def _find_entry(
```

Comment:
I don't see any advantage to combining the functions find-by-id and find-by-path. What happens if someone supplies all arguments and they conflict?  The way you have coded it find-by-id would win, but there is no warning issued.  Better to avoid the problem altogether by keeping the old `_find_previous` function and adding a new function, e.g. `_find_by_id`, which only accepts an id argument.
You say in the docstring "only one dataset should ever by found", but that's not true if the routine is called using relative_path, owner_type and owner arguments. There can be multiple entries in your result set if the dataset has been overwritten.   Old entries will still be in the db, marked as overwritten. There could be more than one. Under normal circumstances, all but one _should_ be labeled as overwritten.   In the original _find_previous routine my thinking was that if there had been some glitch (e.g., bug in earlier version of code or database failure in a previous register operation before that field could be updated) causing there to be more than one old entry not marked as overwritten that it might as well be fixed this time around.

Reply:
I have split back into two functions. There is now a `find_entry` method in the base table class, which can search for an entry based on an ID (can be used for any table). And `find_previous` as before, which finds all datasets with the combination of `relative_path`, `owner`, and `owner_type` and sees if the latest entry is overwritable.